pkg/operator/disruptioncontroller/disruptioncontroller.go

```diff
@@ -0,0 +1,125 @@
+package disruptioncontroller
+
+import (
+	"context"
+	"errors"
+	"sync"
+	"time"
+
+	operatorv1 "github.com/openshift/api/operator/v1"
+	"github.com/openshift/library-go/pkg/controller/factory"
+	"github.com/openshift/library-go/pkg/operator/events"
+	"github.com/openshift/library-go/pkg/operator/v1helpers"
+	"go.etcd.io/etcd/etcdserver/etcdserverpb"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/connectivity"
+	"k8s.io/klog/v2"
+
+	"github.com/openshift/cluster-etcd-operator/pkg/etcdcli"
+)
+
+const (
+	syncDuration  = 2 * time.Minute
+	retryDuration = 1 * time.Second
+)
+
+// DisruptionController observes the operand for disruption and events observations.
+type DisruptionController struct {
+	operatorClient v1helpers.OperatorClient
+	etcdClient     etcdcli.EtcdClient
+}
+
+func NewDisruptionController(
+	operatorClient v1helpers.OperatorClient,
+	etcdClient etcdcli.EtcdClient,
+	eventRecorder events.Recorder,
+) factory.Controller {
+	c := &DisruptionController{
+		operatorClient: operatorClient,
+		etcdClient:     etcdClient,
+	}
+	return factory.New().ResyncEvery(syncDuration).WithInformers(
+		operatorClient.Informer(),
+	).WithSync(c.sync).ToController("DisruptionController", eventRecorder.WithComponentSuffix("disruption-controller"))
+}
+
+func (c *DisruptionController) sync(ctx context.Context, syncCtx factory.SyncContext) error {
+	err := c.checkDisruption(ctx, syncCtx.Recorder())
+	if err != nil && !errors.Is(err, context.DeadlineExceeded) {
```

Comment:
Is DeadlineExceeded not an error condition?

Reply:
in this context, it is expected[1] but more of a hack for not having proper error handling of the timeout itself.

[1] https://github.com/openshift/cluster-etcd-operator/pull/559/files#diff-3b8f033e644d1031d01f1b25c8c19640c7b8b84d3382508dd83f69c6a567a3a8R72

Reply:
to your point, other client timeouts can exist.

Reply:
If you see here timeout, though, is the intention that the disruption controller not be indicated as degraded? The current condition is binary (degraded/not degraded) and I wonder if it shouldn't be degraded/not degraded/unknown.

Reply:
if we have a client timeout[1] that is a degraded condition if the context timeout[2] which is intended to resync fires then that is expected. 

[1] https://github.com/openshift/cluster-etcd-operator/pull/559/files#diff-3b8f033e644d1031d01f1b25c8c19640c7b8b84d3382508dd83f69c6a567a3a8R76

[2] https://github.com/openshift/cluster-etcd-operator/pull/559/files#diff-3b8f033e644d1031d01f1b25c8c19640c7b8b84d3382508dd83f69c6a567a3a8R71